keps/sig-node/20190425-node-readiness-gates.md

```diff
@@ -0,0 +1,389 @@
+    // +optional
+    Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+    // Whether a node is ready based on it's corresponding readiness gates
+    // +optional
+    Ready bool `json:"ready" protobuf:"varint,2,opt,name=ready"`
+}
+```
+
+Updates for `node.Status.Readiness` will be done by the Node Lifecycle Controller.
+
+### Limitations and Risks
+
+#### Tolerating the Readiness Taint
+
+One limitation in this design is that all system critical pods that should gate the readiness of a node must be updated to tolerate the readiness taint. Otherwise,
+those pods cannot be scheduled on a node since the readiness taint would already be applied by the kubelet on registration. If users forget to apply this toleration,
+they may accidentally gate the readiness of their nodes, preventing workloads from being scheduled there.
+
+#### Unintended Pod Label Query for a Readiness Gate
+
+If a Kubernetes user accidentally uses a label query that matches more pods than they thought, then node readiness may behave in a way they did not expect.
+In the worse case scenario, if too many pods are selected for node readiness, the readiness for a node may flap as non-critical pods are rescheduled across the cluster.
+
+### Graduation Criteria
+
+#### Alpha
+
+- the readiness taint is properly applied to a node if any of its readiness gate pods are not ready
+- all functionality for readiness gates is gated by a feature gate (default off)
+
+#### Alpha -> Beta Graduation
+
+TODO: after alpha
+
+#### Beta -> GA Graduation
+
+TODO: after beta
+
+## Implementation History
+
+TODO: once accepted and KEP status is set to `implementable`.
+
+## Alternatives
+
+### [Approach A] Tracking state of extensions via APIs
+
+For reasons similar to recent [node heartbeats revamp](/kep/sig-node/0009-node-heartbeat.md), individual node extensions will be expected to post their state via dedicated "[Lease](https://github.com/kubernetes/api/blob/master/coordination/v1beta1/types.go#L27)" API objects.
+
+The Node API will include a new field that will track a list of node extensions that are expected on the node.
+This list can be initially populated by the node controller when a node object is created or by the kubelet as it’s extensions register, or by individual extensions themselves (if they have sufficient privileges).
+
+Here are some sample illustrations:
+
+```yaml
+type Node struct {
+ …
+ Spec:
+  SystemExtensions []SystemExtension
+...
+}
+
+type SystemExtension struct {
+   Reference ObjectReference
+   AffectsReadiness bool // Ignore non-critical node extensions from node readiness. Helps identify the overall set of extensions deterministically.
+}
+```
+
+The following journey attempts to illustrate this solution a bit more in detail:
+
+1. Kubelet starts up on a node and immediately taints the node with a special “kubernetes.io/MetaReadiness” taint with an effect of “NoSchedule” in addition to flipping it’s “Ready” condition to true (when appropriate).
```

Comment:
Maybe "ApplicationReadiness" rather than "MetaReadiness"?

Reply:
and not NodeReadiness?